---
title: "Evaluating optimised genomics workflows"
image: ../figs/Parabricks.png
---

As a part of the Australian BioCommons [Bring Your Own Data platforms project](https://www.biocommons.org.au/byo-data-platform-expansion), we work closely with partners from the [National Computational Infrastructure (NCI)](https://nci.org.au/), who were looking to make scalable bioinformatics data processing more accessible at their facilities. We worked with NCI to evaluate the performance of NVIDIA Parabricks software suite at their Gadi HPC. GPU-accelerated tools, like Parabricks, are currently rare in bioinformatics but have the capacity to significantly reduce the time and computational resources required for complex data processing like mapping and variant calling. 

#### My role

* Design and implement a technical performance evaluation protocol 
* Design and implement a biological validation protocol 
* Run and compare the performance of multiple optimised genomic workflows at NCI Gadi 
* Report findings and recommendations to NCI, NVIDIA, and Australian BioCommons regarding the potential impact of the availability of Parabricks at NCI 

#### Outcomes

* Workflow: [GermlineShortV Biovalidation](https://github.com/Sydney-Informatics-Hub/GermlineShortV_biovalidation)
* Report: [Evaluation of Mapping and Germline Variant Calling Pipelines on Australian High-Performance Computing Facilities](https://zenodo.org/record/6930814)
* Access to Parabricks licenses at NCI Gadi at no cost to researchers

#### Skills 

* Data analysis, project management, verbal and written communication
* Disciplines: bioinformatics, genomics 
* Platforms: HPC  
* Tools: bash, PBS, git, GitHub
